### ISSUE TITLE
Proposal: easily configure redirect path when unauthenticated

### ISSUE THREAD

**Issue by kylefox** [plataformatec/devise#4921]
By default Devise will redirect to `"new_#{scope}_session_url"` when the `authenticate_#{scope}!` filter fails.

There are a number of StackOverflow Q&As floating around that suggest overriding `authenticate_user!` to redirect to the page you want. But this is not a good approach because it bypasses a lot of the stuff Devise does behind the scenes, such as redirecting back to the attempted URL after sign in.

I wrote a wiki article on how to do this correctly by defining a custom failure app with a `route` method that returns your desired route:

- **Wiki:** [Redirect to new registration (sign up) path if unauthenticated](https://github.com/plataformatec/devise/wiki/Redirect-to-new-registration-(sign-up)-path-if-unauthenticated)

However, it would be _much_ nicer if this was a configurable option in the initializer, maybe something like this:

```ruby
config.authentication_failure_route = :new_user_registration_path
```

It should also allow a lambda, so that a different route could be returned depending on what type of scope it is. For example, if you have both and `Admin` and `User`, you don't want `authenticated_admin!` redirecting to `new_admin_registration_path`:

```ruby
config.authentication_failure_route = lambda do |scope|
  scope.to_sym == :user ? :new_user_registration_url : :"new_#{scope}_session_url"
end
```

Would the Devise team consider adding this functionality? I can prepare a PR if so.

**Comment by tegon** [comment]
Seems good, that would definitely help people trying to customize this behavior.

I think we should keep the `route` method so that people that are overriding this like your Wiki article wouldn't need to change anything.

Also, we might think better on what exactly provide to the block. In your example, it's only `scope` - which might be enough for most use cases - but we might also include other Warden stuff that can be useful (maybe `attempted_path` and `message`, or even expose the whole `warden_options` hash). I'm not sure about this though, so I'd love to hear more thoughts on it (yours and from other people in the community).

Thoughts? @feliperenan @rafaelfranca

**Comment by rafaelfranca** [comment]
I'm fine with this new option too.

**Comment by tegon** [comment]
@kylefox Are you going to work on this? I'm asking so that people know whether this is up for grabs or not.

**Comment by kylefox** [comment]
@tegon No, I don't think I'll have a chance to get to this any time soon 😢

You (or anyone else) should definitely feel free to work on it!

**Comment by tinyfive** [comment]
If no one is working on this, I would like to take it.

**Comment by tegon** [comment]
@tinyfive Thanks, feel free do it.
